### ISSUE TITLE
Regression tests for model outputs

### ISSUE THREAD

**Issue by joeloskarsson** [mllam/neural-lam#59]
Something that would be really nice to have is regression testing for model outputs. In short, whenever we refactor something in models we want them to still be able to load checkpoints (or well, see #48 ) and give exactly the same output when being fed with the same data.

One way to achieve this could be to

1. Check out main branch
2. Run some example data through the model and save the predictions (potentially also some internal representation tensors but likely unneccesary and hard to do in practice)
3. Check out PR
4. Run the same example data through the model and compare outputs to saved predictions.

I'm not too familiar with `pytest` and the github workflows to know all the details of how to do this. @SimonKamuk, @leifdenby  do you think something like this is doable? Or are there any better ways to achieve this?

**Comment by Sharkyii** [comment]
@joeloskarsson, I would like to work on this. My idea is to implement a CI workflow (.github/workflows/test.yml) that automatically validates model checkpoints whenever a Pull Request is opened.

The workflow would function as follows:

Baseline Generation: We generate and store a output (baseline predictions) from the current main branch.
PR Validation: When a PR is submitted, the workflow runs the new code against the same input data.
Comparison: It compares the new outputs against the baseline. If the outputs differ beyond a strict floating-point tolerance (e.g., atol=1e-5), the test fails.

Should i proceed with the mentioned idea?

**Comment by Saptami191** [comment]
Hi @joeloskarsson and @Sharkyii,

I think adding regression tests for model outputs would be very valuable,after studying the docs and working with a small isuue i thought this. one concern I have with dynamically generating the baseline from main inside CI is reproducibility small differences in environment (CPU vs GPU, PyTorch versions, floating-point behavior) could lead to false failures even if the model hasn’t meaningfully changed.
An alternative could be to generate a fixed reference output once (using CPU inference with a fixed seed and small example dataset) and store it in the repository. CI would then compare model outputs against this frozen reference with a strict tolerance. This keeps the test deterministic and avoids relying on cross-branch comparisons during CI.
Would this approach be simpler and more stable long-term? I’d be happy to prototype a minimal version.
see my approach is **simpler CI workflow** that is needed  i think so,**version-controlled baseline**. So i will go ahead with that issue if maintainers give a green button on my approach. I have worked much on regression test and forecasting model.

**Comment by sahilkr31** [comment]
Hi,
I’ve been reading through the discussion on regression testing for model outputs.
The fixed reference approach (CPU inference with a fixed seed and a small example dataset) seems more stable long-term than dynamically comparing against main in CI, especially considering potential floating-point or environment differences.
I’d be interested in prototyping a minimal deterministic regression test using:

- Fixed seed
- CPU inference
- A small example dataset
- Strict tolerance comparison

Before starting, I just wanted to clarify:

1. Should the reference outputs be stored in the repository?
2. Is there already a preferred minimal dataset for inference testing?

I’d be happy to draft a small initial implementation for feedback.

**Comment by joeloskarsson** [comment]
Some good discussion going on here 😄  Here's my thoughts:

1. I don't think non-determinism should be a big issue here. We do not really need to run this on GPU, where most non-determinism issues arise. We should always use float-comparisons with reasonable tolerances.
2. If I understand correctly, the two alternatives seems to be either 1. running both the baseline (before PR) and new (after merged PR) models within the test, and comparing their outputs, or 2. keeping the output of the baseline model stored somewhere and then comparing to this. Alternative 1 seems simpler, but also requires more compute within the test.
3. If we should store reference outputs then it should not be in the repo I think, but somewhere externally. Could maybe use the same aws storage that we use for the test datasets for this, but not sure. Would have to check with @leifdenby in that case.
4. I think we can use the same example data as is used for running the other tests for this. Maybe we should have some actually trained checkpoint to use, but that can easily be swapped out at a later point if we have a functioning setup. In my view we just need to run inference for a couple of samples, not check a whole dataset. If something is wrong this should be clear pretty quickly.

**Comment by sahilkr31** [comment]
Hi @joeloskarsson,

Thanks for the detailed explanation — that helps a lot.

Based on your suggestions, I’ll proceed with a minimal regression test that runs on CPU, uses a small number of samples from the existing example data, and compares outputs using float tolerances.
I’ll keep it simple and avoid storing large reference outputs. The idea is to run inference on a few samples so any unintended change becomes visible quickly.
I’ll open a draft PR once I have something working so we can iterate on it.

**Comment by Saptami191** [comment]
@sahilkr31 I am already working on this issue.

**Comment by sahilkr31** [comment]
@Saptami191 Thanks for the heads up.

I wasn’t aware you were already working on it. 
Happy to coordinate — let me know if you’d prefer I wait for your PR 
or if there’s a part where collaboration would be helpful.

Looking forward to contributing.

**Comment by Saptami191** [comment]
@joeloskarsson 
this PR adds a deterministic regression test for GraphLAM to prevent unintended numerical changes during refactoring.the baseline is version-controlled for simplicity and CI stability, but I’m happy to adjust the storage approach if preferred.
